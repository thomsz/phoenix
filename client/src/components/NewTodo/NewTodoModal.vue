<template>
  <div class="flex flex-col space-y-3 absolute w-full bg-white/80 backdrop-blur-sm top-16">
    <hr />
    <textarea
      name="description"
      rows="5"
      class="w-full p-2 bg-transparent border-slate-100 border-b-2"
      v-model="form.description"
      placeholder="Descriptionâ€¦"
    >
    </textarea>
    <label class="px-2 text-gray-400 hover:cursor-pointer">
      Due date:
      <input
        name="due_date"
        type="date"
        class="h-8 bg-transparent"
        v-model="form.dueDate"
      />
    </label>
    <div class="w-full flex gap-2">
      <button
        class="action bg-blue-100 hover:bg-blue-200 grow"
        :class="{ 'bg-blue-400 hover:bg-blue-500': form.title }"
      >
        Create task
      </button>
      <button
        type="button"
        class="action w-12 hover:bg-gray-300"
        @click="$emit('close-button-clicked')"
      >
        <XIcon class="close icon fill-slate-300" />
      </button>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import { XIcon } from '@heroicons/vue/solid'

export default defineComponent({
  name: 'NewTodoModal',

  components: {
    XIcon
  },

  props: {
    modelValue: {
      type: Object,
      required: true
    }
  },

  data () {
    return {
      form: this.modelValue
    }
  },

  watch: {
    form: {
      deep: true,
      handler (form): void {
        this.$emit('update:modelValue', form)
      }
    }
  }
})
</script>

<style scoped>
button.action {
  @apply flex justify-center items-center h-12 px-3 rounded-md font-bold uppercase text-white transition-colors
    duration-500;
}

button.action:hover .icon.close {
  @apply fill-white;
}
</style>
