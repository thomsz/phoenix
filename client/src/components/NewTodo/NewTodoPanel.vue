<template>
  <form
    ref="newTodo"
    name="new_todo"
    @submit.prevent="createTodo"
  >
    <input
      name="title"
      v-model="form.title"
      required
    />
    <div>
      <input
        name="description"
        v-model="form.description"
      />
      <input
        name="due_date"
        type="date"
        v-model="form.dueDate"
      />
      <button>Add</button>
    </div>
  </form>
</template>

<script lang="ts">
import { v4 as uuid } from 'uuid'
import { defineComponent } from 'vue'

import type Todo from '@/types/Todo'

export default defineComponent({
  name: 'NewTodoPanel',

  data: () => ({
    form: {
      title: '',
      dueDate: '',
      description: ''
    }
  }),

  methods: {
    resetForm (): void {
      // TODO: typing
      (this.$refs.newTodo as any).reset()
    },

    createTodo (): void {
      // TODO: create todo
      const { title, dueDate, description } = this.form
      const todo: Todo = {
        id: uuid(),
        title,
        description,
        ...(dueDate && { dueDate: new Date(dueDate) }),
      }

      this.resetForm()
      this.$emit('todo-created', todo)
    }
  }
})
</script>
