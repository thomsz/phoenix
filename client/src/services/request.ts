import env from '@/services/env'
import request from 'axios'
import type { AxiosInstance } from 'axios'

const generateRequestInstance = (): AxiosInstance => request.create({
  baseURL: env.API_URL
})

export default generateRequestInstance()
