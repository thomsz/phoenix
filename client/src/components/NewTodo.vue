 <template>
  <form
    name="new_todo"
    @submit.prevent="createTodo"
  >
    <div class="flex">
      <input
        name="title"
        class="w-full h-16 text-2xl px-2"
        v-model="form.title"
        required
        placeholder="Add a new taskâ€¦"
        @focus="expanded = true"
      />
      <ReturnIcon
        v-if="expanded"
        class="w-10 mx-2 fill-slate-200 hidden md:block"
      />
    </div>
    <Transition name="fade">
      <NewTodoModal
        v-if="expanded"
        v-model="form"
        @close-button-clicked="expanded = false"
      />
    </Transition>
  </form>
</template>

<script lang="ts">
import request from '@/services/request'
import { defineComponent } from 'vue'

import ReturnIcon from '@/components/icons/ReturnIcon.vue'
import NewTodoModal from '@/components/NewTodo/NewTodoModal.vue'

import type Todo from '@/types/Todo'

const generateForm = () => ({
  title: '',
  dueDate: '',
  description: ''
})

export default defineComponent({
  name: 'NewTodo',

  components: {
    ReturnIcon,
    NewTodoModal
  },

  data: () => ({
    expanded: false,
    form: generateForm()
  }),

  watch: {
    'form.title' (title: string): void {
      if (title) {
        this.expanded = true
      }
    }
  },

  methods: {
    resetForm (): void {
      this.expanded = false
      this.form = generateForm()
    },

    async createTodo (): Promise<void> {
      const { title, dueDate, description } = this.form
      const item: Omit<Todo, 'id'> = {
        done: false,
        title,
        dueDate: dueDate ? new Date(dueDate).toISOString() : '',
        description
      }

      try {
        const response = await request.post('tasks', item)
        const createdItem: Todo = response.data
        this.resetForm()
        this.$emit('todo-created', createdItem)
      } catch (error) {
        console.error(error)
      }

    }
  }
})
</script>

<style scoped>
form {
  @apply flex flex-col relative z-10;
}

.fade-enter-active {
  @apply duration-300 ease-out;
}

.fade-enter-from {
  @apply opacity-0;
}

.fade-enter-to {
  @apply opacity-100;
}

.fade-leave-active {
  @apply duration-200 ease-in;
}

.fade-leave-from {
  @apply opacity-100;
}

.fade-leave-to {
  @apply opacity-0;
}
</style>
