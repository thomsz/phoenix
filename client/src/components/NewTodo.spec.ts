import request from '@/services/request'
import factory from '@/tests/services/factory'
import { shallowMount } from '@vue/test-utils'

import NewTodo from '@/components/NewTodo.vue'
import ReturnIcon from '@/components/icons/ReturnIcon.vue'
import NewTodoModal from '@/components/NewTodo/NewTodoModal.vue'

import type Todo from '@/types/Todo'

jest.mock('@/services/env', () => ({}))

describe('NewTodo', () => {
  it('should create an item', async () => {
    const fakeItem: Todo = {
      ...factory.generate('Todo'),
      done: false
    }

    jest.spyOn(request, 'post').mockImplementation((url: string) => {
      switch (url) {
        case 'tasks':
          return Promise.resolve({
            data: fakeItem
          })

        default:
          return Promise.reject()
      }
    })

    const wrapper = shallowMount(NewTodo)
    expect(wrapper.findComponent(ReturnIcon).exists()).toBe(false)
    expect(wrapper.findComponent(NewTodoModal).exists()).toBe(false)

    await wrapper.find('input').trigger('focus')
    expect(wrapper.findComponent(ReturnIcon).exists()).toBe(true)
    expect(wrapper.findComponent(NewTodoModal).exists()).toBe(true)

    wrapper.find('input').setValue(fakeItem.title)
    wrapper.findComponent(NewTodoModal).setValue(fakeItem)
    await wrapper.find('form').trigger('submit')
    expect(request.post).toBeCalledWith('tasks', {
      done: false,
      title: fakeItem.title,
      dueDate: fakeItem.dueDate,
      description: fakeItem.description
    })

    await wrapper.vm.$nextTick()
    expect(wrapper.findComponent(ReturnIcon).exists()).toBe(false)
    expect(wrapper.emitted('todo-created')?.[0][0]).toEqual(fakeItem)
  })

  it('should close modal', async () => {
    const wrapper = shallowMount(NewTodo, {
      data: () => ({
        expanded: true
      })
    })

    expect(wrapper.html()).toMatchSnapshot()
    const newTodoModalWrapper = wrapper.findComponent(NewTodoModal)
    await newTodoModalWrapper.vm.$emit('close-button-clicked')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
