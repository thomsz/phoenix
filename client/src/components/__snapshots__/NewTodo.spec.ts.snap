// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTodo should close modal 1`] = `
"<form name=\\"new_todo\\" class=\\"flex flex-col relative z-10\\">
  <div class=\\"flex\\"><input name=\\"title\\" class=\\"w-full h-16 text-2xl px-2\\" required=\\"\\" placeholder=\\"Add a new task…\\">
    <return-icon-stub class=\\"w-10 mx-2 fill-slate-200 hidden md:block\\"></return-icon-stub>
  </div>
  <transition-stub>
    <new-todo-modal-stub modelvalue=\\"[object Object]\\"></new-todo-modal-stub>
  </transition-stub>
</form>"
`;

exports[`NewTodo should close modal 2`] = `
"<form name=\\"new_todo\\" class=\\"flex flex-col relative z-10\\">
  <div class=\\"flex\\"><input name=\\"title\\" class=\\"w-full h-16 text-2xl px-2\\" required=\\"\\" placeholder=\\"Add a new task…\\">
    <!--v-if-->
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</form>"
`;
